src/main/java/fi/aalto/cs/apluscourses/intellij/model/CommonLibraryProvider.java

```diff
@@ -0,0 +1,43 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import fi.aalto.cs.apluscourses.model.Component;
+import fi.aalto.cs.apluscourses.model.ComponentSource;
+import fi.aalto.cs.apluscourses.model.Library;
+import fi.aalto.cs.apluscourses.model.NoSuchComponentException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+class CommonLibraryProvider implements ComponentSource {
```

Comment:
What is a common library?

Reply:
A library that is not provided by the course but it is assumed that the plugin knows where to get it. Our only library `ScalaSdk` is a common library.

Generally, if a course-wide library (which is not a module like O1Lib but binary) is included in the project, the course configuration file should hold the information (URL) how to get it (we don't currently have support to this, but it is taken into account in design). Some libraries, i.e. common libraries, are though handled separately: it is enough that they are referenced in IML files, and our plugin knows where to get them. Scala SDK is (so far only) this kind of library.

Reply:
Makes sense, thanks for the explanation! And is the `Library` class the common base class for both course libraries and common libraries? If so, should it also have a URL?

Reply:
Yes, `Library` is a base class but... as common libraries "do not exist" in course configuration file there is not common/course library separation outside the IntelliJ specific part of our code.

`IntelliJLibrary` is also a base class for both common and course libraries; it is there in-between. (See the UML chart made by Nikolai. It is also included in this PR but it can be only viewed with IDEA Ultimate.)

It is not yet designed (as it is not really needed in O1) how course libraries are represented in course configuration file (one URL or many URLs or something else?) so I wouldn't add URL field to `Library` class. Also, it might be that in future there are common libraries whose binaries are e.g. shipped with our plugin so they wouldn't have a URL at all. For these reasons, I would keep URL out of the base class.

![Annotation_2020-04-15_154026](https://user-images.githubusercontent.com/6066364/79338303-bb40d480-7f2f-11ea-90ef-4f805ccaa1fb.png)

Reply:
Okay, sounds good :+1: By the way great chart @superseacat!

Reply:
thanks, but it's auto-generated, take a look at: [IntelliJ IDEA's manual](https://www.jetbrains.com/help/idea/class-diagram.html) (the main challenge is to pick the right scope to display) :stuck_out_tongue: